units/en/unit2/hands-on.mdx

```diff
@@ -0,0 +1,1074 @@
+
+In this notebook, **you'll code from scratch your first Reinforcement Learning agent** playing FrozenLake ❄️ using Q-Learning, share it to the community, and experiment with different configurations.
+
+
+⬇️ Here is an example of what **you will achieve in just a couple of minutes.** ⬇️
+
+
+<img src="https://huggingface.co/datasets/huggingface-deep-rl-course/course-images/resolve/main/en/unit3/envs.gif" alt="Environments"/>
+
+### 🎮 Environments:
+
+- [FrozenLake-v1](https://www.gymlibrary.dev/environments/toy_text/frozen_lake/)
+- [Taxi-v3](https://www.gymlibrary.dev/environments/toy_text/taxi/)
+
+### 📚 RL-Library:
+
+- Python and Numpy
+
+We're constantly trying to improve our tutorials, so **if you find some issues in this notebook**, please [open an issue on the Github Repo](https://github.com/huggingface/deep-rl-class/issues).
+
+## Objectives of this notebook 🏆
+
+At the end of the notebook, you will:
+
+- Be able to use **Gym**, the environment library.
+- Be able to code from scratch a Q-Learning agent.
+- Be able to **push your trained agent and the code to the Hub** with a nice video replay and an evaluation score 🔥.
+
+
+## Prerequisites 🏗️
+
+Before diving into the notebook, you need to:
+
+🔲 📚 **Study [Q-Learning by reading Unit 2](https://huggingface.co/deep-rl-course/unit2/introduction)**  🤗
+
+
+## Install dependencies and create a virtual display 🔽
+
+During the notebook, we'll need to generate a replay video. To do so, with colab, **we need to have a virtual screen to be able to render the environment** (and thus record the frames).
+
+Hence the following cell will install the librairies and create and run a virtual screen 🖥
+
+We’ll install multiple ones:
+
+- `gym`: Contains the FrozenLake-v1 ⛄ and Taxi-v3 🚕 environments. We use `gym==0.24` since it contains a nice Taxi-v3 UI version.
+- `pygame`: Used for the FrozenLake-v1 and Taxi-v3 UI.
+- `numPy`: Used for handling our Q-table.
+
+The Hugging Face Hub 🤗 works as a central place where anyone can share and explore models and datasets. It has versioning, metrics, visualizations and other features that will allow you to easily collaborate with others.
+
+
+You can see here all the Deep reinforcement Learning models available 👉 https://huggingface.co/models?other=q-learning
+
+
+```bash
+pip install -r https://github.com/huggingface/deep-rl-class/tree/main/notebooks/unit2/requirements-unit2.txt
+```
+
+```bash
+sudo apt-get update
+apt install python-opengl
+apt install ffmpeg
+apt install xvfb
+pip3 install pyvirtualdisplay
+```
+
+To make sure the new installed libraries are used, **sometimes it's required to restart the notebook runtime**. The next cell will force the **runtime to crash, so you'll need to connect again and run the code starting from here**. Thanks for this trick, **we will be able to run our virtual screen.**
+
+```python
+import os
+
+os.kill(os.getpid(), 9)
```

Comment:
FYI I just tried without restarting and it worked well (would be good to confirm)

Reply:
Same, but given some might experience this problem I was thinking of keeping it :thinking:

Reply:
(for now)